pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	klog.Infof("processing key %q", key)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	err := c.process(ctx, key)
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		klog.Infof("Successfully reconciled %q", key)
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
+	num := c.queue.NumRequeues(key)
+	if errors.IsRetryable(err) || num < 5 {
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Errorf("Dropping key %q after failed retries: %v", key, err)
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	namespace, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != "" {
+		return fmt.Errorf("namespace %q found in key for cluster-wide Workspace object", namespace)
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.workspaceLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		return err
+	}
+	previous := obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Spec, obj.Spec) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().Update(ctx, obj, metav1.UpdateOptions{})
+		return uerr
+	}
+	if !equality.Semantic.DeepEqual(previous.Status, obj.Status) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().UpdateStatus(ctx, obj, metav1.UpdateOptions{})
+		return uerr
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, obj *tenancyv1alpha1.Workspace) error {
+	if currentShard := obj.Status.Location.Current; currentShard != "" {
+		// make sure current shard still exists
+		shards, err := c.workspaceShardLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		var found bool
+		for _, shard := range shards {
+			if shard.Name == currentShard {
+				found = true
+				break
+			}
+		}
+		if !found {
+			klog.Infof("removing spec.targetShard for workspace %q", obj.Name)
+			obj.Spec.TargetShard = ""
```

Comment:
a controller should not do this. It is the intent of the creator of the object. Don't override that. If it is not possible to "apply" that, tell the user in a condition.

Reply:
I can see that you set the target if it has been empty before. It's also what kube-scheduler does. But already that is a questionable behaviour, but one can argue that it is like defaulting basically.

Reply:
We need a different design (`WorkspaceBinding`?) if we don't want to mutate spec.

Reply:
setting it to "" is strange. Setting "" to some value is more natural. Why do we need this "unschedule" case?

Reply:
I just wanted something sensible to handle shard movement for now, is all